Client/Frontend/Sync/BraveCore/BraveCoreMigrator.swift

```diff
@@ -261,7 +267,136 @@ class BraveCoreMigrator {
     }
 }
 
+// MARK: - History Migration
+
+extension BraveCoreMigrator {
+    
+    private func migrateHistoryModels(_ completion: @escaping (Bool) -> Void) {
+        if !Preferences.Chromium.syncV2HistoryMigrationCompleted.value {
+            // If the history model has already loaded, the observer does NOT get called!
+            // Therefore we should continue to migrate the history
+            if historyAPI?.isBackendLoaded == true {
+                performHistoryMigrationIfNeeded { success in
+                    completion(success)
+                }
+            } else {
+                // Wait for the history service to load before we attempt to perform migration!
+                self.historyObserver = historyAPI?.add(HistoryServiceLoadedObserver({ [weak self] in
+                    guard let self = self else { return }
+                    self.historyObserver?.destroy()
+                    self.historyObserver = nil
+
+                    self.performHistoryMigrationIfNeeded { success in
+                        completion(success)
+                    }
+                }))            }
+        } else {
+            completion(true)
+        }
+    }
+    
+    private func performHistoryMigrationIfNeeded(_ completion: ((Bool) -> Void)?) {
+        log.info("Migrating to Chromium History v1 - Start")
+        migrateHistory() { success in
+            Preferences.Chromium.syncV2HistoryMigrationCompleted.value = success
+            completion?(success)
+        }
+    }
+    
+    private func migrateHistory(_ completion: @escaping (_ success: Bool) -> Void) {
+        DataController.performOnMainContext { context in
```

Comment:
I wonder if this will be a performance problem, imagine users using Brave daily for 5 years, the history DB will be huge right

Reply:
It will not create a problem cause only the visual portion we show to user which is last month's history is migrated.
Please check comment  https://github.com/brave/brave-ios/pull/3703#discussion_r677619942

Also addHistory method from brave-core has to be called in main thread.